Move Global Styles APIs to @wordpress/block-editor

noisysocks opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
- Moves `useSetting` in `edit-site` → `useGlobalSetting` in `block-editor`.
- Moves `useStyle` in `edit-site` → `useGlobalStyle` in `block-editor`.
- Moves `useGlobalStylesReset`, `useGlobalStylesOutput` and `GlobalStylesContext` in `edit-site` to `block-editor` as is.
- `lock()`s the above APIs within `wp.blockEditor.experiments`.
- The UI components and code that references the `global-styles` CPT remain in `@wordpress/edit-site`.

## Why?
More and more there is a need for blocks to access Global Styles. For example we want to:

- Show to the user what the default value will be (e.g. in a `placeholder`) is if a control is left empty ([context](https://github.com/WordPress/gutenberg/pull/46894))
- Display recently created style variations in the block inspector ([context](https://github.com/WordPress/gutenberg/pull/46343))
- Allow editing global styles from block inspector using _Push to Global Styles_ button ([context](https://github.com/WordPress/gutenberg/pull/46446))

https://github.com/WordPress/gutenberg/pull/46894 and https://github.com/WordPress/gutenberg/pull/34178 address this need by adding `useStyle` (to compliment `useSetting`) to `@wordpress/block-editor`. It's awkward though because any implementation of `useStyle` needs `getValueFromVariable` to make sense of CSS variables. This is an implementation detail of and not something that I think should be exposed from `@wordpress/block-editor`.

Instead we can move the high level Global Styles APIs to `@wordpress/block-editor` and make Global Styles a first class citizen of the block editor. This seems inevitable to me and to some extent has already begun. For example the block editor has a `__experimentalFeatures` setting which contains `mergedConfig.settings` and the block settings REST API has `__experimentalStyles` which contains `mergedConfig.styles`.

## Testing Instructions
There's no modified or new functionality here so just need to check for regressions. 

1. Go through each of the panels in Appearance → Editor → Styles and verify that there are no errors.
2. Test modifying and saving a few different global styles.
3. Test that 'Push globally' button still works by modifying the styles of a block and pressing 'Push globally' in the 'Advanced' panel.

noisysocks committed:
Move things around

noisysocks commented:
The outstanding questions I'm thinking about are:

1) Can we merge the existing `useSetting` in `@wordpress/block-editor` with the `useSetting` in `@wordpress/edit-site`?

I am thinking no and have both. One is a higher level `useSetting( path )` function that looks up the block from context and returns the specified setting with inheritance. The other is a lower level `useGlobalSetting( path, blockName, source )` function that only looks at the Global Styles config object but lets you change it by returning a setter. Same for styles.

2) Do we need both `GlobalStylesContext` and `BlockEditorProvider`?

I think no, since having both creates two sources of truth which is confusing. I'm not sure of the best API though. Do we simply add `baseConfig`, `userConfig`, and `setUserConfig` to the block editor settings and deprecate `__experimentalFeatures`? I'm not sure that such an API would be clear to non-WordPress consumers of `@wordpress/block-editor` as `config` is a pretty meaningless word.

noisysocks commented:
@oandregal @youknowriad: What are your thoughts on the proposed architecture here and my questions above? (Ignore the code in this PR which is rough.) While researching `useSetting` and `__experimentalFeatures` I came across a lot of discussion between the two of you on such things.

oandregal commented:
> Can we merge the existing useSetting in @wordpress/block-editor with the useSetting in @wordpress/edit-site?

Yeah, what you said. The root difference between the two is that block editor's useSettings inspects _block instances_, while edit site's useSetting inspects _block types_.

I remember that there was discussion about this at some point, and people mentioned that it could be worth having a global styles package, to support use cases such as onboarding, for example. I don't have a strong opinion on being a standalone package or part of block-editor package. One of those makes sense.

> Do we need both GlobalStylesContext and BlockEditorProvider?

Depending on whether global styles it's a standalone package or part of the block-editor, I guess the answer would be different?

cc @jorgefilipecosta as I think he'd have thoughts on the topic as well.

noisysocks commented:
Is it worth creating a new package when `@wordpress/block-editor` would have to depend on it anyway? In other words it wouldn't decrease the bundle size of `block-editor.js` or anything like that.

tellthemachines commented:
> Can we merge the existing useSetting in @wordpress/block-editor with the useSetting in @wordpress/edit-site?

Might be better to keep them separate for the reasons @oandregal mentions above. I do think that whatever we do here should be reflected in `useStyle` as they are very similar in structure.

> Is it worth creating a new package when @wordpress/block-editor would have to depend on it anyway? In other words it wouldn't decrease the bundle size of block-editor.js or anything like that.

Probably not 😅 I'd be happy for it to live in block-editor.

youknowriad commented:
> Can we merge the existing useSetting in @wordpress/block-editor with the useSetting in @wordpress/edit-site?

I think one could use the other (the one with clientId can use the other one) under the hood and augment it with "instance specific" logic. But that might require a lot of work, so fine to post pone probably.

> Do we need both GlobalStylesContext and BlockEditorProvider?

Not entirely certain, the global styles variations use-case might force us to reply "yes" here unless we find a way to do it with the latter.

> baseConfig, userConfig, and setUserConfig to the block editor settings and deprecate __experimentalFeatures? I

I agree that there's some duplication here, not clear thought what's the best solution :)

---
There's another area where there is duplication that we should ideally remove: the UI. The different block inspector controls and the different levels of the global styles UI are very similar and sometimes we have duplicated components. We've had a discussion on how to [unify these](https://github.com/WordPress/gutenberg/discussions/37064). Probably a lot of work for an unclear outcome.

The elephant in the room here is that if this folder moves into the block-editor package, it means a lot of new stable APIs to maintain for a piece of UI/behavior that still in flux (though stabilizing more and more). This might mean, it's better to start with a "bundled" `@wordpress/global-styles` package instead (even without the core-data dependency) as this allows us to iterate on it as we wish without third-party impact.

noisysocks commented:
> The elephant in the room here is that if this folder moves into the block-editor package, it means a lot of new stable APIs to maintain for a piece of UI/behavior that still in flux (though stabilizing more and more). This might mean, it's better to start with a "bundled" `@wordpress/global-styles` package instead (even without the core-data dependency) as this allows us to iterate on it as we wish without third-party impact.

That's a really compelling point!

noisysocks committed:
lock() the Global Styles APIs

noisysocks committed:
Update package-lock.json

noisysocks committed:
Move tests to block-editor

noisysocks commented:
I tried to split the global styles APIs into a new "bundled" `@wordpress/global-styles` package as suggested above but quickly ran into a problem: doing this means that there are circular dependencies between `block-editor` and `global-styles` as `global-styles` depends on [these `block-editor` APIs](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-site/src/components/global-styles/use-global-styles-output.js#L18-L22) and `block-editor` depends on `global-styles` in oder to implement `useSetting`, `useStyle`, etc.

Since they're tightly coupled, I moved the global styles APIs back to `block-editor` but `lock()`ed them within an `experiments` object so that they cannot be used by third parties for now.

I left `GlobalStylesContext` alone for now. We can probably integrate it into `BlockEditorProvider` but best to do this separately so that this PR is only concerned with moving around / renaming existing APIs.

I renamed `useStyle` and `useSetting` to `useGlobalStyle` and `useGlobalSetting` so that there is no confusing them with the identically named hooks that use block context.

In a follow-up PR we can refactor `useSetting` to use `useGlobalSetting` internally and add a new `useStyle` hook that uses `useGlobalStyle` internally. 

---

So, to clarify, what this PR does is:

- Moves `useSetting` in `edit-site` → `useGlobalSetting` in `block-editor`.
- Moves `useStyle` in `edit-site` → `useGlobalStyle` in `block-editor`.
- Moves `useGlobalStylesReset`, `useGlobalStylesOutput` and `GlobalStylesContext` in `edit-site` to `block-editor` as is.
- `lock()`s the above APIs within `wp.blockEditor.experiments`.

youknowriad commented:
>  block-editor depends on global-styles in oder to implement useSetting, useStyle, etc.

You could also argue that my suggestion here was rubbish and just avoid this dependency (like today)

youknowriad commented on packages/block-editor/src/components/global-styles/index.js:
```diff
@@ -0,0 +1,7 @@
+	useGlobalStyle,
+} from './hooks';
+export { useGlobalStylesOutput } from './use-global-styles-output';
+export { GlobalStylesContext } from './context';
```
Do we want to give the whole context object as an API or just the "provider" for external consumers.

noisysocks replied:
There's a few places where we need to read the raw context value e.g. variations.

https://github.com/WordPress/gutenberg/blob/64c69123ea57a3c342475c8a72d274d4cf9764a3/packages/edit-site/src/components/global-styles/screen-style-variations.js#L45

youknowriad commented on packages/edit-site/src/components/global-styles/README.md:
```diff
@@ -15,77 +15,3 @@ function MyComponent() {
 	return <GlobalStylesUI />;
```
Any thoughts on the UI? can we share with block inspector UI at some point?

noisysocks replied:
It should be possible I think to create high level UI components that live in `@wordpress/block-editor` and are used by both the block inspector and global styles. Ideally only the `value` and `onChange` differs. So basically what you propose in https://github.com/WordPress/gutenberg/discussions/37064. I note that since you opened that discussion we've come a long way in making the UIs look very similar so ideally now we can make the shared components be very high level (e.g. `TypographyPanel`, `DimensionsPanel`).

I'll make a note to explore this!

youknowriad reviewed (approved):
Looks good to me, personally, I still have some uncertainties but this looks like a good start. 

Btw, I'd love to be pinged on PRs when we'll use these APIs in other parts of the block editor.

noisysocks commented:
> You could also argue that my suggestion here was rubbish and just avoid this dependency (like today)

Even if we didn't make `useStyle` use `useGlobalStyle` internally we'd still need `block-editor` to depend on `global-styles` because there's features that we want to implement that would use `useGlobalStyle` e.g. the push to global styles button.

> I still have some uncertainties

Yeah, me too 😄

> Btw, I'd love to be pinged on PRs when we'll use these APIs in other parts of the block editor.

Will keep you posted!
